### DISCUSSION TITLE
have git smartlog only show the last N days/hours of branches/commit

### DISCUSSION THREAD

**Discussion by dashed** [arxanas/git-branchless#140] [Ideas]
`git sl` can sometimes be too verbose and show everything. 

It'll be nice if `git sl` can be smarter, and only display the last N days/hours of branches/commit. This would be an alternative to having to `git hide`/`git unhide` commits. 

Any thoughts on this proposal?

**Comment by arxanas** [comment]
The `hg sl` command I used to use hid commits that were older than two weeks, which is probably worth implementing.

> This would be an alternative to having to git hide/git unhide commits.

I'd like to understand more about your workflow. In my workflow, I typically `git move` old commit stacks onto the main branch, which will automatically get rid of any commits which have been applied upstream. You can expect a `git sync` command to be implemented sometime soon to make it easier to bring all of your work up-to-date (and get rid of any upstream-merged work).

How many branches are you using/do you use any long-running branches which appear in the smartlog? Currently, there's no way to "abbreviate" a segment of commits which aren't on the main branch. So if your hope is to hide most commits on a feature branch, then it would be more complicated than just hiding old commit stacks.

**Reply by arxanas** [reply]
> You can expect a git sync command to be implemented sometime soon to make it easier to bring all of your work up-to-date (and get rid of any upstream-merged work).

[`git sync`](https://github.com/arxanas/git-branchless/wiki/Command:-git-sync) is now implemented.

> Currently, there's no way to "abbreviate" a segment of commits which aren't on the main branch

`git smartlog` now supports abbreviation. If your [revset](https://github.com/arxanas/git-branchless/wiki/Command:-git-smartlog#custom-queries) doesn't include a given commit, but includes ancestor and descendant commits, then it will insert an abbreviated section where topologically necessary. For example, just rendering the branches abbreviates several commits for me:

```
$ git sl 'branches()'
⋮
◇ 5cffad2 22d docs: update `README.md`
┣━┓
⋮ ◯ 6ea3a82 22d (arxanas/amend) temp: add --interactive flag to `git amend`
⋮
◇ b7d4952 15d (master) fix(amend): don't detach current branch when amending
┣━┓
┃ ⊘ 2 omitted commits
┃ ⋮
┃ ◯ ef64436 3d (arxanas/snapshot) temp: fix some compilation errors?
┃ ⋮
┃ ⊘ 1 omitted descendant commit
┣━┓
┃ ⊘ 1 omitted commit
┃ ⋮
┃ ● 717890d 1d (ᐅ arxanas/build) build: run `cargo update`
⋮
⊘ 2 omitted commits
⋮
◯ 043af38 3d (arxanas/record) temp: attempt to optimize UI
```

**Comment by dashed** [comment]
> I'd like to understand more about your workflow. 

I primarily use Sourcetree (https://www.sourcetreeapp.com) for most of the git operations I use. I still use the command line for more surgical git rebasing or cherry-pick operations. 

Since operations like `git move`, `git prev`, and `git next` operate on a detached `HEAD`, `git sl` has been useful in visualizing the tree. For comparison, Sourcetree is unable to work against a detached `HEAD`. 

As an aside, Sourcetree doesn't seem to be able to detect the pre-commit hooks that `git branchless init` installs. I've yet to investigate further to fixing it.

Other tools I've used are https://github.com/Shopify/git-chain. I've written my own version of `git-chain` (resides in https://github.com/dashed/git-chain); which I'm using right now. This is very similar to `git restack`. 

> How many branches are you using/do you use any long-running branches which appear in the smartlog?

I have long-standing branches for changes that I've not yet committed to deleting. In my specific case, there were commits that were ~61 days old showing up in `git sl`. 

For context, I was trialing `git branchless` for work (https://github.com/getsentry/sentry); which is where my stale branches show up.

**Reply by epage** [reply]
> Since operations like git move, git prev, and git next operate on a detached HEAD, git sl has been useful in visualizing the tree. For comparison, Sourcetree is unable to work against a detached HEAD.
>
> As an aside, Sourcetree doesn't seem to be able to detect the pre-commit hooks that git branchless init installs. I've yet to investigate further to fixing it.

Sounds like [git-stack](https://github.com/epage/git-stack) might be a better fit, at least for now.  It uses an attached HEAD and does not rely on hooks.

> Other tools I've used are https://github.com/Shopify/git-chain. I've written my own version of git-chain (resides in https://github.com/dashed/git-chain); which I'm using right now. This is very similar to git restack.

Oooh, more tools to add to my [stacking comparison](https://github.com/epage/git-stack/blob/main/docs/comparison.md)

**Reply by dashed** [reply]
> Sounds like [git-stack](https://github.com/epage/git-stack) might be a better fit, at least for now. It uses an attached HEAD and does not rely on hooks.

I was aware if your implementation of `git-stack` and as well as other tools. At the time I was shopping for a "git stack" tool, Shopify's `git-chain` seemed to resonate well to me for my pressing needs. 

I'll definitely try out `git-stack`! 👍

**Comment by arxanas** [comment]
This is now possible by passing a revset like `committer.date('after: 7 days ago')` to `git smartlog`. See the [revset reference](https://github.com/arxanas/git-branchless/wiki/Reference:-Revsets) for more information. You might have to tweak the revset to get what you want; you can start by modifying the smartlog's [default revset](https://github.com/arxanas/git-branchless/wiki/Command:-git-smartlog#default-revset).
